File: run_backend.py
import os
import subprocess
import sys

root_dir = os.path.dirname(os.path.abspath(__file__))

# Check if a local virtualenv exists
venv_python_win = os.path.join(root_dir, ".venv", "Scripts", "python.exe")
venv_python_posix = os.path.join(root_dir, ".venv", "bin", "python")

if os.path.exists(venv_python_win):
    py_bin = venv_python_win
elif os.path.exists(venv_python_posix):
    py_bin = venv_python_posix
else:
    py_bin = sys.executable

cmd = [
    py_bin,
    "-m",
    "uvicorn",
    "backend.api.main:app",
    "--reload",
    "--host",
    "127.0.0.1",
    "--port",
    "8000",
    "--app-dir",
    root_dir,
]

sys.exit(subprocess.call(cmd))

File: run-backend.js
const { spawn, execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const rootDir = path.resolve(__dirname);

function getPython() {
  const venvWin = path.join(rootDir, '.venv', 'Scripts', 'python.exe');
  const venvPosix = path.join(rootDir, '.venv', 'bin', 'python');

  if (fs.existsSync(venvWin)) return venvWin;
  if (fs.existsSync(venvPosix)) return venvPosix;

  if (process.platform === 'win32') {
    try {
      execSync('py -0', { stdio: 'ignore' });
      return 'py';
    } catch {
      return 'python';
    }
  }

  return 'python3';
}

const pyBin = getPython();
const runBackendPy = path.join(rootDir, 'run_backend.py');

const child = spawn(pyBin, [runBackendPy], {
  cwd: rootDir,
  stdio: 'inherit',
  shell: true,
});

child.on('exit', (code) => {
  process.exit(code || 0);
});

child.on('error', (err) => {
  console.error('Failed to start backend server:', err);
  process.exit(1);
});
